### PR TITLE
TextField: new InternalTextField for reusability

### PR DESCRIPTION
**Pull request by AlbertCarreras**
TextField: new internal InternalTextField for reusability

This PR is a preparatory step for upcoming [ComboBox PR](https://github.com/pinterest/gestalt/pull/1563) which shares an `InternalTextField` with TextField.  `InternalTextField` extends the TextField's API.

Some of this logic _could_ be elevated to TextField itself but as a first step, we are deduping the logic first, keeping sam API for TextField

What InternalTextField supports that TextField doesn't:
- implements internal `ref` changes if tags are passed so Popover can get correctly positioned when tags are passed (otherwise Textfields with helper texts or error messages break the spacing between TextField and Popover)
- adds Interactive Icons at the end of the Textfield to clear values or show/expand Popover. Implements `accessibilityClearButtonLabel` for the icons
- adds new `onClick` prop to capture click events 
- adds new `onClickIconButton` prop to as callback functions to execute after clicking new icon button 
- implements `aria-activedescendant` and `accessibilityActiveDescendant` props to support `combobox`, lists, and `options`
- implements `accessibilityControls` to support `combobox`,
-    
### Links

- [Jira](link to Jira ticket(s))
- [TDD](link to Paper doc)
- [Figma](link to Figma file)

### Checklist

- [ ] Added unit and Flow Tests
- [ ] Added documentation + accessibility tests
- [ ] Verified accessibility: keyboard & screen reader interaction
- [ ] Checked dark mode, responsiveness, and right-to-left support
- [ ] Checked stakeholder feedback (e.g. Gestalt designers)

### CODE REVIEW

File: packages/gestalt/src/InternalTextField.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  // REQUIRED
+  id: string,
+  onChange: ({|
+    event: SyntheticInputEvent<HTMLInputElement>,
+    value: string,
+  |}) => void,
+  // OPTIONAL
+  accessibilityClearButtonLabel?: string,
+  accessibilityControls?: string,
+  accessibilityActivedescendant?: string,
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js:30]: DONE

File: packages/gestalt/src/InternalTextField.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+  accessibilityShowButtonLabel?: string,
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js:31]: DONE

File: packages/gestalt/src/InternalTextField.js [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+  autoComplete?: 'current-password' | 'new-password' | 'on' | 'off' | 'username' | 'email',
+  disabled?: boolean,
+  endButton?: Button,
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js:34]: `textfieldIconButton`!

[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js:34]: DONE

File: packages/gestalt/src/InternalTextField.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                onBlur={() => setFocusedButton(false)}
+                onFocus={() => setFocusedButton(true)}
+                onKeyDown={({ event }) => {
+                  if (event.keyCode !== ENTER && event.keyCode !== TAB) innerRef.current?.focus();
+                }}
+                onMouseEnter={() => setFocusedButton(true)}
+                onMouseLeave={() => setFocusedButton(false)}
+                onTap={handleOnClickIconButton}
+                rounding="circle"
+                tabIndex={endButton === 'clear' ? 0 : -1}
```
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js:224]: So yes. At some point on my refactor, there was  moment when it was possible to focus on it. But it shouldn't get focus at any time. I'l remove it, cos yes, it doesn't make sense to add that extra logic

[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js]: DONE

File: packages/gestalt/src/InternalTextField.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js]: REMOVED

File: packages/gestalt/src/InternalTextField.js [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+import FormErrorMessage from './FormErrorMessage.js';
+import FormHelperText from './FormHelperText.js';
+import FormLabel from './FormLabel.js';
+import Tag from './Tag.js';
+import layout from './Layout.css';
+import styles from './InternalTextField.css';
+import { TAB, ENTER } from './keyCodes.js';
+import typography from './Typography.css';
+
+type Button = 'clear' | 'dropdown';
```
Comment (ayeshakmaz) [inline packages/gestalt/src/InternalTextField.js:19]: Nit: don't think this needs to be pulled out

**Review by ayeshakmaz — APPROVED** [review]
Tiniest nit comment, but it looks great!

File: packages/gestalt/src/InternalTextField.js (hunk 5 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/InternalTextField.js]: DONE

**Pull request merged**